metrics-exporter-prometheus/src/exporter/push_gateway.rs

```diff
@@ -0,0 +1,136 @@
+use std::time::Duration;
+
+use hyper::{header::HeaderValue, Uri};
+use tracing::error;
+
+use super::ExporterFuture;
+use crate::PrometheusHandle;
+
+// Creates an ExporterFuture implementing a push gateway.
+pub(super) fn new_push_gateway(
+    endpoint: Uri,
+    interval: Duration,
+    username: Option<String>,
+    password: Option<String>,
+    handle: PrometheusHandle,
+) -> ExporterFuture {
+    use http_body_util::{BodyExt, Collected, Full};
+    use hyper::body::{Buf, Bytes};
+    use hyper::{Method, Request};
+    use hyper_tls::HttpsConnector;
+    use hyper_util::{client::legacy::Client, rt::TokioExecutor};
+
+    Box::pin(async move {
+        let https = HttpsConnector::new();
+        let client: Client<_, Full<Bytes>> = Client::builder(TokioExecutor::new())
+            .pool_idle_timeout(Duration::from_secs(30))
+            .build(https);
+
+        let auth = username.as_ref().map(|name| basic_auth(name, password.as_deref()));
+
+        loop {
+            // Sleep for `interval` amount of time, and then do a push.
+            tokio::time::sleep(interval).await;
+
+            let mut builder = Request::builder();
+            if let Some(auth) = &auth {
+                builder = builder.header("authorization", auth.clone());
+            }
+
+            let output = handle.render();
+            let result =
+                builder.method(Method::PUT).uri(endpoint.clone()).body(Full::from(output));
+            let req = match result {
+                Ok(req) => req,
+                Err(e) => {
+                    error!("failed to build push gateway request: {}", e);
+                    continue;
+                }
+            };
+
+            match client.request(req).await {
+                Ok(response) => {
+                    if !response.status().is_success() {
+                        let status = response.status();
+                        let status =
+                            status.canonical_reason().unwrap_or_else(|| status.as_str());
+                        let body = response
+                            .into_body()
+                            .collect()
+                            .await
+                            .map(Collected::aggregate)
+                            .map_err(|_| ())
+                            .map(|mut b| b.copy_to_bytes(b.remaining()))
```

Comment:
Might be a little simpler to use `Collected::to_bytes` above and then you could drop this line entirely, I think.